### PR TITLE
[data] Add `Dataset.mix()` release test microbenchmark

### PR DESCRIPTION
**Pull request by justinvyu**
## Description

Adds a release test benchmark for `Dataset.mix()` (introduced in #63168) that measures mixing throughput and ratio accuracy.

## Benchmark details

Benchmark design:
- Creates 8 datasets reading ImageNet parquet, each stamped with a ds_index column
- Mixes with Dataset.mix(), repartitions to 4 * batch_size rows per block
- Consumes via TorchTrainer to mimic the seen weighting ratio when ingesting multiple local batches which are split across workers.
- Tracks per-batch mixing ratios per worker, aggregates mean/std across workers via all_reduce to get the mean and standard deviation across **global batches.**
- Asserts ratio mean is within 0.05 of target and std < 0.1
- Tests with and without a shuffling step after mixing with `--num-workers=1` to showcase the effectiveness of shuffling removing the dependency of mixing quality on the number of workers. [See here for more details.](https://docs.ray.io/en/master/data/mixing-data.html#random-mixing)

Test variations:
  | Variation | Workers | Weights | Random mix |
  |---|---|---|---|
  | `8ds_equal` | 8 | equal | no |
  | `8ds_power_law` | 8 | 128:64:32:16:8:4:2:1 | no |
  | `8ds_equal_random_mix` | 1 | equal | yes |
  | `8ds_power_law_random_mix` | 1 | 128:64:32:16:8:4:2:1 | yes |

All variants use `--max-rows-per-worker=100_000` for bounded runtime on an 8-node m5.2xlarge cluster, and also to avoid hitting dataset exhaustion which leads to the ratios diverging.

## Benchmark results

8ds_equal
```
Dataset 0: mean=0.1253±0.0695, target=0.1250, diff=0.0003 OK
Dataset 1: mean=0.1252±0.0634, target=0.1250, diff=0.0002 OK
Dataset 2: mean=0.1264±0.0584, target=0.1250, diff=0.0014 OK
Dataset 3: mean=0.1240±0.0600, target=0.1250, diff=0.0010 OK
Dataset 4: mean=0.1254±0.0654, target=0.1250, diff=0.0004 OK
Dataset 5: mean=0.1239±0.0653, target=0.1250, diff=0.0011 OK
Dataset 6: mean=0.1247±0.0630, target=0.1250, diff=0.0003 OK
Dataset 7: mean=0.1250±0.0663, target=0.1250, diff=0.0000 OK
```

8ds_power_law
```
Dataset 0: mean=0.5017±0.0941, target=0.5020, diff=0.0002 OK
Dataset 1: mean=0.2518±0.0858, target=0.2510, diff=0.0008 OK
Dataset 2: mean=0.1244±0.0681, target=0.1255, diff=0.0011 OK
Dataset 3: mean=0.0630±0.0464, target=0.0627, diff=0.0003 OK
Dataset 4: mean=0.0310±0.0327, target=0.0314, diff=0.0004 OK
Dataset 5: mean=0.0160±0.0268, target=0.0157, diff=0.0003 OK
Dataset 6: mean=0.0079±0.0196, target=0.0078, diff=0.0000 OK
Dataset 7: mean=0.0042±0.0151, target=0.0039, diff=0.0003 OK
```

8ds_equal_random_mix
```
  main = {'time': 55.192007467999986, 'object_store_spilled_total_gb': 0.0, 'num_datasets': 8, 'weights': None, 'num_workers': 1, 'batch_size': 256, 'max_rows_per_worker': 100000, 'stopping_condition': 'stop_on_longest_drop', 'random_mix': True, 'print_every': 50, 'tput': 2453.388865730177, 'num_rows': 100096, 'ratio_mean_ds0': 0.12647858056265984, 'ratio_std_ds0': 0.08976139552706039, 'ratio_mean_ds1': 0.12640864769820973, 'ratio_std_ds1': 0.03321393841700443, 'ratio_mean_ds2': 0.12865648976982097, 'ratio_std_ds2': 0.11139429644615093, 'ratio_mean_ds3': 0.12791719948849106, 'ratio_std_ds3': 0.06092154319651775, 'ratio_mean_ds4': 0.12861652813299232, 'ratio_std_ds4': 0.04257910528838874, 'ratio_mean_ds5': 0.12067415281329923, 'ratio_std_ds5': 0.03243481089665854, 'ratio_mean_ds6': 0.12062420076726342, 'ratio_std_ds6': 0.031844610741798195, 'ratio_mean_ds7': 0.12062420076726342, 'ratio_std_ds7': 0.032520041872085376}
```

8ds_power_law_random_mix
```
Dataset 0: mean=0.4944±0.0890, target=0.5020, diff=0.0076 OK
Dataset 1: mean=0.2497±0.0918, target=0.2510, diff=0.0013 OK
Dataset 2: mean=0.1244±0.0953, target=0.1255, diff=0.0011 OK
Dataset 3: mean=0.0606±0.0230, target=0.0627, diff=0.0022 OK
Dataset 4: mean=0.0307±0.0327, target=0.0314, diff=0.0007 OK
Dataset 5: mean=0.0197±0.0302, target=0.0157, diff=0.0040 OK
Dataset 6: mean=0.0102±0.0238, target=0.0078, diff=0.0024 OK
Dataset 7: mean=0.0102±0.0239, target=0.0039, diff=0.0063 OK
```

### CODE REVIEW

**Comment by justinvyu** [comment]
Release tests passed: https://buildkite.com/ray-project/release/builds/92743#_

File: release/nightly_tests/dataset/dataset_mixing/mix_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+                ratio_stdevs = []
+                for i in range(num_ds):
+                    count_sums = torch.tensor(count_history[i], dtype=torch.double)
+                    dist.all_reduce(count_sums, op=dist.ReduceOp.SUM)
+
+                    ratios = count_sums.numpy() / batch_size_sums.numpy()
+                    ratio_means.append(np.mean(ratios))
+                    ratio_stdevs.append(np.std(ratios))
+
+                return ratio_means, ratio_stdevs
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply release/nightly_tests/dataset/dataset_mixing/mix_benchmark.py:121]: equal=True streaming split prevent this

**Review by xinyuangui2 — APPROVED** [review]
